timeutil/schedule.go

```diff
@@ -215,3 +395,341 @@ func ParseSchedule(scheduleSpec string) ([]*Schedule, error) {
+// mon,wed  (Monday and on Wednesday)
+// mon..wed (same as above)
+//
+// Returns a slice of schedules or an error if parsing failed
+func ParseScheduleV2(scheduleSpec string) ([]*Schedule, error) {
+	var schedule []*Schedule
+
+	for _, s := range strings.Split(scheduleSpec, "..") {
+		// cut the schedule in event sets
+		//     eventlist = eventset *( ".." eventset )
+		sched, err := parseEventSetScheduleV2(s)
+		if err != nil {
+			return nil, err
+		}
+		schedule = append(schedule, sched...)
+	}
+	return schedule, nil
+}
+
+// makeWeekdayEvent generates a per day (span) schedule
+func makeWeekdayEvent(start string, end string) (*Schedule, error) {
+	startDay, startMonthWeekDay, _ := parseWeekdaySpec(start)
+	endDay, endMonthWeekDay, _ := parseWeekdaySpec(end)
+
+	if endMonthWeekDay < startMonthWeekDay {
+		return nil, fmt.Errorf("unsupported schedule")
+	}
+
+	sched := &Schedule{
+		Weekday: startDay,
+		// WeekdayEnd:      endDay,
+		MonthWeekday: startMonthWeekDay,
+		// MonthWeekdayEnd: endMonthWeekDay,
+	}
+	if startDay != endDay {
+		sched.WeekdayEnd = endDay
+	}
+	if endMonthWeekDay != startMonthWeekDay {
+		sched.MonthWeekdayEnd = endMonthWeekDay
+	}
+
+	return sched, nil
+}
+
+// parseSpan splits a span string `<start>-<end>` and returns the start and end
+// pieces. If string is not a span then the whole string is returned
+func parseSpan(spec string) (string, string) {
+	var start, end string
+
+	specs := strings.Split(spec, spanToken)
+
+	if len(specs) == 1 {
+		start = specs[0]
+		end = start
+	} else {
+		start = specs[0]
+		end = specs[1]
+	}
+	return start, end
+}
+
+// parseTimeV2 parses a time specification which can either be `<hh>:<mm>` or
+// `<hh>:<mm>-<hh>:<mm>`. Returns corresponding TimeOfDay structs.
+func parseTimeV2(start, end string) (TimeOfDay, TimeOfDay, error) {
+	// is it a time?
+	var err error
+	var tstart, tend TimeOfDay
+
+	if start == end {
+		// single time, eg. 10:00
+		tstart, err = ParseTime(start)
+		if err == nil {
+			tend = tstart
+		}
+	} else {
+		// time span, eg. 10:00-12:00
+		tstart, tend, err = parseTimeInterval(fmt.Sprintf("%s-%s", start, end))
+	}
+
+	return tstart, tend, err
+}
+
+// timeScheduleConf is a helper type for passing current timespec information
+type timeScheduleConf struct {
+	// event start time
+	start TimeOfDay
+	// event end time
+	end TimeOfDay
+	// how many events between start and end, 0 or 1 mean exactly one event
+	count uint
+	// should event time be randomized
+	randomize bool
+}
+
+// makeTimeSchedule creates a single
+func makeTimeSchedule(conf timeScheduleConf, base *Schedule) *Schedule {
+	if base == nil {
+		base = &Schedule{}
+	}
+	nsch := *base
+	nsch.Start = conf.start
+	nsch.End = conf.end
+	nsch.Randomize = conf.randomize
+	return &nsch
+}
+
+// parseWeekdaySpec will parse a string and extract weekday (eg. wed),
+// MonthWeekday (eg. 1st, 5th in the month).
+func parseWeekdaySpec(s string) (string, MonthWeekday, error) {
```

Comment:
It feels better not to have a type for MonthWeekday. This number is going to be seen and considered as a count in most cases. It's literally spelled out in the syntax.. mon3 is 3. So it's natural to represent it as a count internally as well. Just a plain int.